modules/storage/src/StockStorageAPI.php

```diff
@@ -21,180 +21,200 @@ public function createTransaction($variation_id, $location_id, $zone, $quantity,
+      ->fields('it', ['location_id'])
+      ->condition('location_id', $location_id)
+      ->condition('variation_id', $variation_id)
+      ->condition('trid', $min, '>');
+    if ($max) {
+      $query->condition('trid', $max, '<=');
     }
+    $query->addExpression('SUM(qty)', 'qty');
+    $query->groupBy('location_id');
+    $result = $query->execute()
+      ->fetch();
+
+    return $result ? $result->qty : 0;
   }
 
   /**
    * {@inheritdoc}
    */
-  public function getStockLevel($variation_id, array $locations) {
-    /** @var array $location_info */
-    $location_info = $this->getStockLocationLevel($variation_id, $locations);
-    // Add the quantities together and return.
-    $qty = 0;
+  public function getTotalStockLevel($variation_id, array $locations) {
+    $location_info = $this->getLocationsStockLevels($variation_id, $locations);
+    $total = 0;
     foreach ($location_info as $location) {
-      $qty += $location['qty'] + $location['transactions_qty'];
+      $total += $location['qty'];
     }
-    return $qty;
+
+    return $total;
   }
 
   /**
-   * Gets the stock location level.
+   * Gets the stock levels for a set of locations.
    *
    * @param int $variation_id
-   *   The product variation ID.
+   *   The product variation id.
+   *
    * @param array $locations
-   *   Array of locations.
+   *   Array of locations ids.
    *
-   * @return array The stock location level.
-   *   The stock location level.
+   * @return array
+   *   Stock level information indexed by location id with these values:
+   *     - 'qty': The quantity.
+   *     - 'last_transaction': The id of the last transaction.
    */
-  public function getStockLocationLevel($variation_id, array $locations) {
-    // An array to hold stock data for the listed locations.
-    $location_info = [];
+  public function getLocationsStockLevels($variation_id, array $locations) {
+    $location_levels = [];
     foreach ($locations as $location_id) {
-      $location_info[$location_id] = [
-        'qty' => 0,
-        'last_transaction' => 0,
-        'transactions_qty' => 0,
+      $location_level = $this->getLocationStockLevel($location_id, $variation_id);
+      $location_levels[$location_id] = [
+        'qty' => $location_level['qty'],
+        'last_transaction' => $location_level['last_transaction'],
       ];
     }
 
-    // Get the location level & last transaction.
-    $db = \Drupal::database();
-    $result = $db->select('cs_inventory_location_level', 'ill')
-      ->fields('ill')
-      ->condition('location_id', $locations, 'IN')
-      ->condition('variation_id', $variation_id)
-      ->execute()
-      ->fetchAll();
-    if ($result) {
-      foreach ($result as $record) {
-        // Location info for retriving transactions.
-        $location_info[$record->location_id]['qty'] = $record->qty;
-        $location_info[$record->location_id]['last_transaction'] = $record->last_transaction_id;
```

Comment:
The last transaction was not the actual last transaction that determined the stock level.  It was the last one where the stock level was calculated / cached.